migrate: add tag to install the canonical static GRUB config

HuijingHei opened the pull request.

cgwalters reviewed (commented):
Thanks for working on this! An initial review pass here

HuijingHei commented on systemd/bootloader-migrate.service:
```diff
@@ -0,0 +1,19 @@
+Type=oneshot
+ExecCondition=/bin/sh -c '[ "$(ostree config --repo=/sysroot/ostree/repo get sysroot.bootloader 2>/dev/null)" != "none" ]'
+ExecStart=bootupctl adopt-and-update
+ExecStart=/usr/bin/bootupctl migrate-static-grub-config --from-tree
```
Looks more sensible to me.

travier commented on Makefile:
```diff
@@ -37,6 +37,7 @@ install-grub-static:
 .PHONY: install-systemd-unit
 install-systemd-unit:
 	install -m 644 -D -t "${DESTDIR}$(PREFIX)/lib/systemd/system/" systemd/bootloader-update.service
+	install -m 644 -D -t "${DESTDIR}$(PREFIX)/lib/systemd/system/" systemd/bootloader-migrate.service
```
We added a drop-in for the existing unit for the Atomic Desktops: https://pagure.io/workstation-ostree-config/blob/main/f/bootupd.yaml#_17

jlebon replied:
Right. In the Atomic Desktop case, we want to "static-ify" a grub2-mkconfig-generated `grub.cfg`. In the CoreOS case, we just want to take the canonical GRUB configs from the rootfs. (The "static-ify the dynamic config" approach could work too in those old CoreOS systems of course, but since we don't care about dual boot in our case, we can afford to just use the canonical static configs instead and bring those nodes back in sync with modern CoreOS.)

jlebon replied:
Hmm, unless I'm missing something, the line we're commenting on here is just installing the unit, not enabling it.

HuijingHei replied:
Yes, by default the unit is disabled.

cgwalters replied:
Hmm. This unit is not just enabling static configs, it's also doing the `adopt-and-update` flow i.e. updating the bootloader as well (which makes sense!).

Wouldn't it be simpler to structure this then as a drop-in for the `bootloader-update.service` unit? I think we'd trigger it on the firstboot path in OCP or so?

HuijingHei commented on systemd/bootloader-migrate.service:
```diff
@@ -0,0 +1,21 @@
+
+[Service]
+Type=oneshot
+ExecCondition=/bin/sh -c '[ "$(ostree config --repo=/sysroot/ostree/repo get sysroot.bootloader 2>/dev/null)" != "none" ]'
```
Agree, we do not need it when change to adopt. Removed, thank you!

HuijingHei replied:
And it is not suitable to run on dual boot, for example on x86_64, after adopt for BIOS, the bootloader will be set to `none`, and will skip `EFI`, this will make the OS unboot if booting EFI.

jlebon replied:
Yes, sorry for misunderstanding (but for understanding my concern!). I thought that the new approach would _not_ update the bootloader unconditionally.

Even though we will soon turn on bootloader updates by default, I don't want to conflate that goal with this bug fix, and especially wouldn't want to do it in 4.19 now.

HuijingHei commented on src/bootupd.rs:
```diff
@@ -702,6 +719,150 @@ pub(crate) fn client_run_migrate_static_grub_config() -> Result<()> {
+                if !grubconfig.exists() {
+                    // Not existing, maybe UEFI
+                    println!("Not found '{}'", grubconfig.display());
+                    if cfg!(target_arch = "powerpc64") {
```
I check 4.3 on ppc, it is already using `bootloader=none`, maybe better to remove the condition?

HuijingHei commented:
Update to `bootupctl adopt-and-update --with-static-config`, welcome review, thank you all!

jlebon commented:
So, combining the two threads here. How about:
- We land `bootupctl adopt-and-update --with-static-config` here without having a unit.
- In OCP 4.19, we add a generator that outputs the migration unit only if `sysroot.bootloader` is `none`. That's nicer because (1) it minimizes as much as possible what runs on unaffected nodes; it won't even show up in the transaction, and (2) it gets rid of the gnarly `ExecCondition`. Accounting for EUS-to-EUS upgrades, we'd need to keep it for 4.20 as well but then could drop it in 4.21.

jlebon commented:
> In OCP 4.19, we add a generator that outputs the migration unit only if `sysroot.bootloader` is not `none`.

That should be I think "only if `sysroot.bootloader` is unset". s390x sets it to `zipl`. And if somehow it's set to some value other than `none` or `zipl`, we should probably just leave the system alone.

HuijingHei commented:
> I guess so, but meh... I'm also fine to just keep it in rhel-coreos-config. That would mean we wouldn't be able to remove it until 4.22 instead of 4.21, but that's fine. It probably make testing the final fix easier.

Could you help to review https://github.com/coreos/rhel-coreos-config/pull/23 ?

HuijingHei commented:
@jlebon I update the logic that will migrate to static grub config if `sysroot.bootloader` is not set instead of checking it is not `none`. Let me know if it is suitable, thank you!

jlebon commented on src/bios.rs:
```diff
@@ -157,10 +159,62 @@ impl Component for Bios {
+        use openat_ext::OpenatDirExt;
+
+        let grub_config_dir = Utf8PathBuf::from(sysroot_path).join("boot/grub2");
+        let grubdir = destdir
```
Hmm, I don't understand why we sometimes prefix `destdir` and sometimes not in the various path maniulations. If we want to assume that `destdir` can be something other than `/`, then I think some of the paths below need to be adjusted?

HuijingHei replied:
Currently the destdir is `/`, and see https://github.com/coreos/bootupd/blob/main/src/bootupd.rs#L518 and [definition](https://github.com/coreos/bootupd/blob/main/src/bootupd.rs#L455).

HuijingHei replied:
Agree it is confusing, update to "boot/grub2" to make it clear.

jlebon commented on src/bios.rs:
```diff
@@ -157,10 +159,62 @@ impl Component for Bios {
+        }
+
+        // Remove the unused file, backuped as /boot/grub2/grub.cfg.bak
+        destdir.remove_file_optional("boot/loader/grub.cfg")?;
```
Presumably, this should instead be removing the target of the symlink instead of hardcoding.

jlebon commented on src/bios.rs:
```diff
@@ -157,10 +159,62 @@ impl Component for Bios {
+
+        // On BIOS, /boot/grub2/grub.cfg is symlink to /boot/loader/grub.cfg,
+        // should backup it to /boot/grub2/grub.cfg.bak
+        if !grubconfig.exists() {
+            anyhow::bail!("Not found '{}'", grubconfig);
+        } else if grubconfig.is_symlink() {
+            let realconfig = grubdir.read_link(grubconfigs::GRUBCONFIG)?;
+            let realconfig =
+                Utf8PathBuf::from_path_buf(realconfig).expect("Path should be valid UTF-8");
+            // Resolve symlink location
+            let mut current_config = grub_config_dir.clone();
+            current_config.push(realconfig);
+            let backup_config = grub_config_dir.join(grubconfigs::BACKUP);
+
+            // Backup the current GRUB config which is hopefully working right now
+            println!(
+                "Creating a backup of the current GRUB config '{}' in '{}'...",
+                current_config, backup_config
+            );
+            std::fs::copy(&current_config, &backup_config)
+                .context("Failed to backup GRUB config")?;
+            // Remove the symlink (on BIOS)
+            grubdir.remove_file_optional(grubconfigs::GRUBCONFIG)?;
+            crate::grubconfigs::install(&destdir, None, true)?;
+        } else {
+            anyhow::bail!("'{}' is not a symlink", grubconfig);
+        }
```
I think one way to reflow this is

```rust
if !grubconfig.exists() {
    bail
} else if !grubconfig.is_symlink() {
    bail
}

// dedent
```

HuijingHei commented on src/efi.rs:
```diff
@@ -282,6 +309,25 @@ impl Component for Efi {
+
+        // Backup current config and install static config
+        if with_static_config {
+            // Installs the static config if the OSTree bootloader is either not set
+            // or is set to a value other than `none`.
+            if let Some(bootloader) = ostreeutil::get_ostree_bootloader()? {
+                if bootloader != "none" {
```
Yes, good catch, forget to update this. Thank you for the review!

jlebon commented on src/bios.rs:
```diff
@@ -157,10 +159,66 @@ impl Component for Bios {
+            anyhow::bail!("Not found '{}'", grub_config);
+        } else if !grub_config.is_symlink() {
+            anyhow::bail!("'{}' is not a symlink", grub_config);
+        } else {
```
Minor/optional: mentioned this in a previous comment; since both paths above return, we don't strictly need this in an else block. That allows us to outdent one level.

HuijingHei replied:
Got it, sorry for the misunderstooding.

jlebon commented on src/efi.rs:
```diff
@@ -245,11 +245,38 @@ impl Component for Efi {
+        println!("Creating a backup of the current GRUB config on EFI");
+        efidir
+            .copy_file(grubconfigs::GRUBCONFIG, grubconfigs::GRUBCONFIG_BACKUP)
+            .context("Backuping grub.cfg")?;
```
Maybe

```suggestion
            .context("Failed to backup GRUB config")?;
```

for consistency with BIOS.

jlebon commented on src/bios.rs:
```diff
@@ -157,10 +159,66 @@ impl Component for Bios {
+            std::fs::copy(&current_config, &backup_config)
+                .context("Failed to backup GRUB config")?;
+            // Remove the symlink
+            grub_config_dir.remove_file(grubconfigs::GRUBCONFIG)?;
```
Hmm, instead of removing the symlink here, can we just let `grubconfigs::install` overwrite it? That removes the window in which there is no GRUB config.

HuijingHei commented:
Refer to  Jonathan's pointer https://gist.github.com/jlebon/7190e57cef4d705582ecdfa745bf1a66, I did testing using 4.1 BIOS/UEFI firmware with [bootloader-migrate.service](https://github.com/coreos/rhel-coreos-config/pull/23), install grub.cfg and bootuuid.cfg successfully.

On BIOS
```
[core@localhost ~]$ journalctl  -u bootloader-migrate.service | cat 
Jun 17 02:26:07 localhost.localdomain systemd[1]: Starting Static GRUB config migration...
Jun 17 02:26:11 localhost.localdomain bootupctl[864]: ostree repo 'sysroot.bootloader' config option is not set yet
Jun 17 02:26:11 localhost.localdomain bootupctl[864]: Creating a backup of the current GRUB config '/boot/grub2/../loader/grub.cfg' in '/boot/grub2/grub.cfg.backup'...
Jun 17 02:26:11 localhost.localdomain bootupctl[864]: Added 01_users.cfg
Jun 17 02:26:11 localhost.localdomain bootupctl[864]: Added 05_ignition.cfg
Jun 17 02:26:11 localhost.localdomain bootupctl[864]: Added 10_blscfg.cfg
Jun 17 02:26:11 localhost.localdomain bootupctl[864]: Added 14_menu_show_once.cfg
Jun 17 02:26:11 localhost.localdomain bootupctl[864]: Added 30_uefi-firmware.cfg
Jun 17 02:26:11 localhost.localdomain bootupctl[864]: Added 41_custom.cfg
Jun 17 02:26:11 localhost.localdomain bootupctl[864]: Installed: grub.cfg
Jun 17 02:26:11 localhost.localdomain bootupctl[864]: Installed: bootuuid.cfg
Jun 17 02:26:11 localhost.localdomain bootupctl[864]: Static GRUB configuration has been adopted successfully.
Jun 17 02:26:11 localhost.localdomain bootupctl[864]: Adopted and updated: BIOS: grub2-tools-1:2.06-104.el9_6.x86_64
Jun 17 02:26:11 localhost.localdomain systemd[1]: Finished Static GRUB config migration.
```

on EFI
```
[root@localhost grub2]# journalctl  -u bootloader-migrate.service | cat 
Jun 17 02:47:04 localhost.localdomain systemd[1]: Starting Static GRUB config migration...
Jun 17 02:47:07 localhost.localdomain bootupctl[886]: ostree repo 'sysroot.bootloader' config option is not set yet
Jun 17 02:47:07 localhost.localdomain bootupctl[886]: Creating a backup of the current GRUB config on EFI
Jun 17 02:47:07 localhost.localdomain bootupctl[886]: Added 01_users.cfg
Jun 17 02:47:07 localhost.localdomain bootupctl[886]: Added 05_ignition.cfg
Jun 17 02:47:07 localhost.localdomain bootupctl[886]: Added 10_blscfg.cfg
Jun 17 02:47:07 localhost.localdomain bootupctl[886]: Added 14_menu_show_once.cfg
Jun 17 02:47:07 localhost.localdomain bootupctl[886]: Added 30_uefi-firmware.cfg
Jun 17 02:47:07 localhost.localdomain bootupctl[886]: Added 41_custom.cfg
Jun 17 02:47:07 localhost.localdomain bootupctl[886]: Installed: grub.cfg
Jun 17 02:47:07 localhost.localdomain bootupctl[886]: Installed: bootuuid.cfg
Jun 17 02:47:07 localhost.localdomain bootupctl[886]: Installed: "redhat/grub.cfg"
Jun 17 02:47:07 localhost.localdomain bootupctl[886]: Installed: "redhat/bootuuid.cfg"
Jun 17 02:47:07 localhost.localdomain bootupctl[886]: Static GRUB configuration has been adopted successfully.
Jun 17 02:47:07 localhost.localdomain bootupctl[886]: Adopted and updated: EFI: grub2-efi-x64-1:2.06-104.el9_6.x86_64,shim-x64-15.8-4.el9_3.x86_64
Jun 17 02:47:07 localhost.localdomain systemd[1]: Finished Static GRUB config migration.
```

travier commented on src/bios.rs:
```diff
@@ -142,10 +144,63 @@ impl Component for Bios {
+        // On BIOS, /boot/grub2/grub.cfg is symlink to /boot/loader/grub.cfg,
+        // backup it to /boot/grub2/grub.cfg.bak
+        if !grub_config.exists() {
+            anyhow::bail!("Not found '{}'", grub_config);
```
```suggestion
            anyhow::bail!("Could not find '{}'", grub_config);
```

travier commented on src/bios.rs:
```diff
@@ -142,10 +144,63 @@ impl Component for Bios {
+        if !grub_config.exists() {
+            anyhow::bail!("Not found '{}'", grub_config);
+        } else if !grub_config.is_symlink() {
+            anyhow::bail!("'{}' is not a symlink", grub_config);
```
Should this be an error? If this is not a symlink then we need to keep going as the migration might have been interrupted after we did symlink swap.

HuijingHei replied:
If `/boot/grub2/grub.cfg` is not symlink, maybe skip backup, then will install / overwrite the `grub.cfg` with static config, and can not remove the config as it is the static config that we need.
What we want to remove is `/boot/grub2/../loader/grub.cfg`, but now it is `/boot/grub2/grub.cfg`. Need to check if it is symlink before `remove_file`, WDYT?

travier replied:
https://github.com/coreos/bootupd/pull/945#discussion_r2156982205

travier commented on src/bios.rs:
```diff
@@ -142,10 +144,63 @@ impl Component for Bios {
+        use anyhow::Context;
+        use openat_ext::OpenatDirExt;
+
+        // sysroot_path is /, destdir is Dir of /
+        let grub_config_path = Utf8PathBuf::from(sysroot_path).join("boot/grub2");
+        let grub_config_dir = destdir
+            .sub_dir("boot/grub2")
+            .context("Opening boot/grub2")?;
```
What the difference between those two paths here?

HuijingHei replied:
`grub_config_path` is `Utf8PathBuf` of `/boot/grub2`, that will be used to backup.
`grub_config_dir` is `Openat::Dir`, that will be used as `read_link` the real path of `/boot/grub2/grub.cfg` and do sync.
So we need both, WDYT?

travier commented on src/bios.rs:
```diff
@@ -142,10 +144,63 @@ impl Component for Bios {
+        );
+        std::fs::copy(&current_config, &backup_config).context("Failed to backup GRUB config")?;
+
+        crate::grubconfigs::install(&destdir, None, true)?;
```
Is this doing an atomic swap with the symlink?

travier replied:
https://docs.rs/openat-ext/latest/openat_ext/trait.OpenatDirExt.html#method.write_file_contents
it should

travier commented on src/efi.rs:
```diff
@@ -246,11 +246,38 @@ impl Component for Efi {
+        efidir
+            .copy_file(grubconfigs::GRUBCONFIG, grubconfigs::GRUBCONFIG_BACKUP)
+            .context("Failed to backup GRUB config")?;
+        grubconfigs::install(&sysroot, Some(&vendor), true)?;
```
Same question, we need to make sure it's doing an atomic swap

travier commented on src/efi.rs:
```diff
@@ -246,11 +246,38 @@ impl Component for Efi {
+        let efidir = destdir
+            .sub_dir(&vendor)
+            .with_context(|| format!("Opening EFI/{}", vendor))?;
+        println!("Creating a backup of the current GRUB config on EFI");
```
Maybe we should not do this if the file already exists (from a previously interrupted migration), but this is an edge case so OK to not check it as well

travier reviewed (commented):
Thanks, mostly LGTM with some questions / things I don't understand / know

HuijingHei commented on src/bios.rs:
```diff
@@ -142,10 +144,63 @@ impl Component for Bios {
+    // - Remove old "/boot/loader/grub.cfg"
+    fn migrate_static_grub_config(&self, sysroot_path: &str, destdir: &openat::Dir) -> Result<()> {
+        use anyhow::Context;
+        use openat_ext::OpenatDirExt;
```
Move to the top now, thank you for the review!

HuijingHei commented:
I did testing on 4.1 BIOS with the current change, after upgrade and reboot, check that:
- create backup
- remove the real config of symlink
- install the static grub configuration
```
[root@localhost core]# journalctl  -u bootloader-migrate.service | cat
Jun 19 05:39:00 localhost.localdomain systemd[1]: Starting Static GRUB config migration...
Jun 19 05:39:03 localhost.localdomain bootupctl[874]: ostree repo 'sysroot.bootloader' config option is not set yet
Jun 19 05:39:03 localhost.localdomain bootupctl[874]: Creating a backup of the current GRUB config '/boot/grub2/../loader/grub.cfg' in '/boot/grub2/grub.cfg.backup'...
...
Jun 19 05:39:03 localhost.localdomain bootupctl[874]: Installed: grub.cfg
Jun 19 05:39:03 localhost.localdomain bootupctl[874]: Installed: bootuuid.cfg
Jun 19 05:39:03 localhost.localdomain bootupctl[874]: Removing /boot/grub2/../loader/grub.cfg
Jun 19 05:39:03 localhost.localdomain bootupctl[874]: Static GRUB configuration has been adopted successfully.
Jun 19 05:39:03 localhost.localdomain bootupctl[874]: Adopted and updated: BIOS: grub2-tools-1:2.06-104.el9_6.x86_64
Jun 19 05:39:03 localhost.localdomain systemd[1]: Finished Static GRUB config migration.
```

In case that process is unfinished, need to rerun. 
Unset `sysroot.bootloader` and remove `/boot/bootupd-state.json`, restart the service again:
- check grub.cfg is not symlink
- backup file existing so skip the backup
- do not remove the config as it is not symlink
- install the static grub configuration

```
ostree config --repo=/sysroot/ostree/repo unset sysroot.bootloader
rm -v /boot/bootupd-state.json 

# systemctl restart bootloader-migrate.service

# journalctl  -u bootloader-migrate.service | cat
Jun 19 05:41:15 localhost.localdomain bootupctl[1219]: ostree repo 'sysroot.bootloader' config option is not set yet
Jun 19 05:41:15 localhost.localdomain bootupctl[1219]: '/boot/grub2/grub.cfg' is not a symlink
...
Jun 19 05:41:15 localhost.localdomain bootupctl[1219]: Installed: grub.cfg
Jun 19 05:41:15 localhost.localdomain bootupctl[1219]: Installed: bootuuid.cfg
Jun 19 05:41:15 localhost.localdomain bootupctl[1219]: Static GRUB configuration has been adopted successfully.
Jun 19 05:41:15 localhost.localdomain bootupctl[1219]: Adopted and updated: BIOS: grub2-tools-1:2.06-104.el9_6.x86_64
Jun 19 05:41:15 localhost.localdomain systemd[1]: Finished Static GRUB config migration.
```

travier commented on src/bios.rs:
```diff
@@ -142,10 +145,69 @@ impl Component for Bios {
+        // Remove the real config if it is symlink
+        if current_config != grub_config {
+            println!("Removing {}", current_config);
+            grub_config_dir.remove_file(current_config.as_std_path())?;
```
This is a clean up and not worth stopping the migration for. We can ignore the error and log it instead.

travier replied:
https://docs.rs/openat-ext/latest/openat_ext/trait.OpenatDirExt.html#tymethod.remove_file_optional

travier replied:
The if condition looks good

travier reviewed (approved):
Two last nits but this LGTM. Thanks a lot
